test/FunctionalTests/Microsoft.OData.Core.Tests/ODataWriterDerivedTypeConstraintTests.cs

```diff
@@ -0,0 +1,735 @@
+            edmCnAddressType.AddStructuralProperty("PostCode", EdmPrimitiveTypeKind.String);
+
+            edmModel.AddElements(new[] { edmAddressType, edmUsAddressType, edmCnAddressType });
+
+            // EntityContainer
+            EdmEntityContainer container = new EdmEntityContainer("NS", "Default");
+            edmMe = container.AddSingleton("Me", edmCustomerType);
+            edmCustomers = container.AddEntitySet("Customers", edmCustomerType);
+            edmModel.AddElement(container);
+
+            // resource
+            odataCustomerResource = new ODataResource
+            {
+                TypeName = "NS.Customer",
+                Properties = new[] { new ODataProperty { Name = "Id", Value = 7 } }
+            };
+
+            odataVipResource = new ODataResource
+            {
+                TypeName = "NS.VipCustomer",
+                Properties = new[] { new ODataProperty { Name = "Id", Value = 8 }, new ODataProperty { Name = "Vip", Value = "Boss" } }
+            };
+
+            odataNormalResource = new ODataResource
+            {
+                TypeName = "NS.NormalCustomer",
+                Properties = new[] { new ODataProperty { Name = "Id", Value = 9 }, new ODataProperty { Name = "Email", Value = "a@abc.com" } }
+            };
+
+            odataAddressResource = new ODataResource
+            {
+                TypeName = "NS.Address",
+                Properties = new[] { new ODataProperty { Name = "Street", Value = "Way" }}
+            };
+
+            odataUsAddressResource = new ODataResource
+            {
+                TypeName = "NS.UsAddress",
+                Properties = new[] { new ODataProperty {Name = "Street", Value = "RedWay"}, new ODataProperty {Name = "ZipCode", Value = 98052}}
+            };
+
+            odataCnAddressResource = new ODataResource
+            {
+                TypeName = "NS.CnAddress",
+                Properties = new[] { new ODataProperty {Name = "Street", Value = "ShaWay"}, new ODataProperty {Name = "PostCode", Value = "201100"}}
+            };
+        }
+
+        #region Singleton Derived Type Constraints
+        [Fact]
+        public void WritingResourceWithSameTypeAsSingletonTypeWithOrWithoutDerivedTypeConstraintWorks()
+        {
+            Action<ODataMessageWriter> writeCustomerAction = (messageWriter) =>
+            {
+                ODataWriter writer = messageWriter.CreateODataResourceWriter(this.edmMe, this.edmCustomerType); // Singleton
+                writer.WriteStart(this.odataCustomerResource);
+                writer.WriteEnd();
+            };
+
+            // Singleton doesn't have the derived type constraint.
+            string customerActual = GetWriterOutput(this.edmModel, writeCustomerAction);
+            Assert.Contains("/$metadata#Me\",\"Id\":7}", customerActual);
+
+            // Singleton has the derived type constraint.
+            SetDerivedTypeAnnotation(this.edmModel, this.edmMe, "NS.VipCustomer");

```

Comment:
>SetDerivedTypeAnnotation [](start = 12, length = 24)

does a new instance of edmModel get created with each test case run, or is the derived type constraint set for all subsequent tests?

Reply:
@mikepizzo For every single test, xUnit.net creates a new instance of the test class, so, in each test method, the "this.edmModel" is the basic edm model created in the constructor. 

Below is the description from xUnit document at: https://xunit.github.io/docs/shared-context

_**xUnit.net creates a new instance of the test class for every test that is run, so any code which is placed into the constructor of the test class will be run for every single test**. This makes the constructor a convenient place to put reusable context setup code where you want to share the code without sharing object instances (meaning, you get a clean copy of the context object(s) for every test that is run)._

Reply:
👍 Thanks for the research @xuzhg!